### PR TITLE
Overhaul of profiler

### LINKED ISSUE #50: Replay block `746218`
The replay of block `746218` reports over 70k calls to `dict_squash`. This means users have defined more than 70k dictionaries which is quite unlikely.

Actions to do:
- replay block with the latest version of `starknet-replay` and check if this anomaly persists
- make a test contract that creates x numbers of dictionaries and verify that `dict_squash` is called x time from the profiler report

### PR DESCRIPTION
**Pull request by Eagle941**
This PR contains the overhaul of the profiler functions to determine the libfunc frequencies.

In addition, more unit tests are added to prove profiler results.

This closes #50

### CODE REVIEW

File: starknet-replay/src/profiler/mod.rs [hunk 1]
```diff
@@ -122,172 +144,97 @@ impl SierraProfiler {
...
+    /// usage.
+    ///
+    /// # Arguments
+    ///
+    /// - `statements`: The map with the frequency of Sierra statements.
+    #[must_use]
+    pub fn unpack_profiling_info(
+        &self,
+        statements: &HashMap<StatementIdx, usize>,
+    ) -> HashMap<SmolStr, usize> {
```
Comment (iamrecursion) [inline starknet-replay/src/profiler/mod.rs:228]: I don't see any reason to be using `SmolStr` in the interface.

**Review by iamrecursion — CHANGES_REQUESTED** [review]
Good stuff, but it needs some tweaks.

File: starknet-replay/src/profiler/analysis.rs [hunk 2]
```diff
@@ -155,40 +137,920 @@ mod tests {
...
-    #[test]
-    fn test_get_profiling_info_processor_params() {
-        // Checking that the important settings are setup correctly to generate
-        // a histogram of libfuncs frequency.
-        let profiling_info_processor_params = get_profiling_info_processor_params();
-        assert_eq!(profiling_info_processor_params.min_weight, 1);
-        assert!(profiling_info_processor_params.process_by_concrete_libfunc);
+    fn read_sierra_compressed_program(filename: &str) -> Program {
+        let sierra_program_json =
+            read_test_file(filename).unwrap_or_else(|_| panic!("Unable to read file {filename}"));
```
[the remark this answers is not in the document]
Comment (Eagle941) [reply starknet-replay/src/profiler/analysis.rs:142]: [Clippy](https://rust-lang.github.io/rust-clippy/master/index.html#/expect_fun_call) forces me to use `unwrap_or_else`

**Review by iamrecursion — APPROVED** [review]
Awesome stuff! The tests are _way_ less magic now.

**Pull request merged**